### DISCUSSION TITLE
azure pipelines scaler not working, using KEDA V2.4

### DISCUSSION THREAD

**Discussion by SandipGorde-TomTom** [kedacore/keda#2332] [Q&A / Need Help]
Hello,

I need a help in finding the exact problem as I do not see any errors or problems while running scaler for azure pipelines. However scaler simply do not work. Below are the more details.

1. Running EKS clsuter
2. deployed KEDA 2.4 version using helm and direct k8s file
3. deployed ado-agent k8s deployment with one agent(pod)
4. deployed scaledobject which deployed hpa 
5. run multiple jobs on the azure pipelines 

Expected behavior: Pipelines agents should scale based on the queue length
Actual behaviors:  pods are not scaling without any error

Using reference guide: https://keda.sh/blog/2021-05-27-azure-pipelines-scaler/

Below is the scaledobject 

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: pipeline-auth
data:
  personalAccessToken: <PAT>
---
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: pipeline-trigger-auth
spec:
  secretTargetRef:
    - parameter: personalAccessToken
      name: pipeline-auth
      key: personalAccessToken
---
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: azure-pipelines-scaledobject
spec:
  scaleTargetRef:
    name: azdevops-deployment
  minReplicaCount: 1
  maxReplicaCount: 5 
  triggers:
  - type: azure-pipelines
    metadata:
      poolID: "1"
      organizationURLFromEnv: "AZP_URL"
    authenticationRef:
     name: pipeline-trigger-auth
```

**Comment by tomkerkhove** [comment]
Can you share the KEDA logs please?

ALso, I presume `AZP_URL` is correctly configured on your deployment?

**Reply by SandipGorde-TomTom** [reply]
```
I1124 13:42:02.367118       1 request.go:655] Throttling request took 1.035939482s, request: GET:https://10.100.0.1:443/apis/monitoring.coreos.com/v1?timeout=32s
2021-11-24T13:42:02.831Z        INFO    controller-runtime.metrics      metrics server is starting to listen    {"addr": ":8080"}
2021-11-24T13:42:02.832Z        INFO    controllers.ScaledObject        Running on Kubernetes 1.21+     {"version": "v1.21.2-eks-0389ca3"}
2021-11-24T13:42:02.833Z        INFO    setup   Starting manager
2021-11-24T13:42:02.833Z        INFO    setup   KEDA Version: 2.4.0
2021-11-24T13:42:02.833Z        INFO    setup   Git Commit:
2021-11-24T13:42:02.833Z        INFO    setup   Go Version: go1.15.13
2021-11-24T13:42:02.833Z        INFO    setup   Go OS/Arch: linux/amd64
I1124 13:42:02.833730       1 leaderelection.go:243] attempting to acquire leader lease keda/operator.keda.sh...
2021-11-24T13:42:02.833Z        INFO    controller-runtime.manager      starting metrics server {"path": "/metrics"}
I1124 13:42:20.238857       1 leaderelection.go:253] successfully acquired lease keda/operator.keda.sh
2021-11-24T13:42:20.239Z        INFO    controller      Starting EventSource    {"reconcilerGroup": "keda.sh", "reconcilerKind": "ClusterTriggerAuthentication", "controller": "clustertriggerauthentication", "source": "kind source: /, Kind="}
2021-11-24T13:42:20.239Z        INFO    controller      Starting EventSource    {"reconcilerGroup": "keda.sh", "reconcilerKind": "ScaledObject", "controller": "scaledobject", "source": "kind source: /, Kind="}
2021-11-24T13:42:20.239Z        INFO    controller      Starting EventSource    {"reconcilerGroup": "keda.sh", "reconcilerKind": "ScaledJob", "controller": "scaledjob", "source": "kind source: /, Kind="}
2021-11-24T13:42:20.239Z        INFO    controller      Starting EventSource    {"reconcilerGroup": "keda.sh", "reconcilerKind": "TriggerAuthentication", "controller": "triggerauthentication", "source": "kind source: /, Kind="}
2021-11-24T13:42:20.339Z        INFO    controller      Starting Controller     {"reconcilerGroup": "keda.sh", "reconcilerKind": "ClusterTriggerAuthentication", "controller": "clustertriggerauthentication"}
2021-11-24T13:42:20.339Z        INFO    controller      Starting workers        {"reconcilerGroup": "keda.sh", "reconcilerKind": "ClusterTriggerAuthentication", "controller": "clustertriggerauthentication", "worker count": 1}
2021-11-24T13:42:20.339Z        INFO    controller      Starting Controller     {"reconcilerGroup": "keda.sh", "reconcilerKind": "ScaledJob", "controller": "scaledjob"}
2021-11-24T13:42:20.339Z        INFO    controller      Starting workers        {"reconcilerGroup": "keda.sh", "reconcilerKind": "ScaledJob", "controller": "scaledjob", "worker count": 1}
2021-11-24T13:42:20.339Z        INFO    controller      Starting Controller     {"reconcilerGroup": "keda.sh", "reconcilerKind": "TriggerAuthentication", "controller": "triggerauthentication"}
2021-11-24T13:42:20.339Z        INFO    controller      Starting EventSource    {"reconcilerGroup": "keda.sh", "reconcilerKind": "ScaledObject", "controller": "scaledobject", "source": "kind source: /, Kind="}
2021-11-24T13:42:20.340Z        INFO    controller      Starting workers        {"reconcilerGroup": "keda.sh", "reconcilerKind": "TriggerAuthentication", "controller": "triggerauthentication", "worker count": 1}
2021-11-24T13:42:20.440Z        INFO    controller      Starting Controller     {"reconcilerGroup": "keda.sh", "reconcilerKind": "ScaledObject", "controller": "scaledobject"}
2021-11-24T13:42:20.440Z        INFO    controller      Starting workers        {"reconcilerGroup": "keda.sh", "reconcilerKind": "ScaledObject", "controller": "scaledobject", "worker count": 1}
2021-11-24T13:42:20.440Z        INFO    controllers.ScaledObject        Reconciling ScaledObject        {"ScaledObject.Namespace": "ado-agent", "ScaledObject.Name": "azure-pipelines-scaledobject"}
2021-11-24T13:42:20.750Z        INFO    controllers.ScaledObject        Initializing Scaling logic according to ScaledObject Specification                                                                                                  {"ScaledObject.Namespace": "ado-agent", "ScaledObject.Name": "azure-pipelines-scaledobject"}
```

**Reply by SandipGorde-TomTom** [reply]
above are the logs of keda operator pod

**Reply by tomkerkhove** [reply]
And there is nothing more? Can you share the info of your scaled object please?

**Reply by tomkerkhove** [reply]
> ALso, I presume AZP_URL is correctly configured on your deployment?

And verify this please.

**Reply by SandipGorde-TomTom** [reply]
> 

yes AZP_URL is correctly mentioned. Organization is the name of our organization from azure devops

https://dev.azure.com/organization

**Reply by SandipGorde-TomTom** [reply]
> 

```
apiVersion: v1
kind: Secret
metadata:
  name: pipeline-auth
data:
  personalAccessToken: <PAT>
---
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: pipeline-trigger-auth
spec:
  secretTargetRef:
    - parameter: personalAccessToken
      name: pipeline-auth
      key: personalAccessToken
---
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: azure-pipelines-scaledobject
spec:
  scaleTargetRef:
    name: azdevops-deployment
  minReplicaCount: 1
  maxReplicaCount: 5 
  triggers:
  - type: azure-pipelines
    metadata:
      poolID: "1"
      organizationURLFromEnv: "AZP_URL"
    authenticationRef:
     name: pipeline-trigger-auth
```

**Reply by tomkerkhove** [reply]
Thanks for sharing but I was more referring to the status of your scaled object in Kubernetes (describe)

**Reply by SandipGorde-TomTom** [reply]
```
Name:         azure-pipelines-scaledobject
Namespace:    ado-agent
Labels:       scaledobject.keda.sh/name=azure-pipelines-scaledobject
Annotations:  <none>
API Version:  keda.sh/v1alpha1
Kind:         ScaledObject
Metadata:
  Creation Timestamp:  2021-11-24T12:49:24Z
  Finalizers:
    finalizer.keda.sh
  Generation:  1
  Managed Fields:
    [...]
  Resource Version:  12687175
  UID:               91276e48-5e4e-472f-97c4-7537a0be615a
Spec:
  Max Replica Count:  5
  Min Replica Count:  1
  Scale Target Ref:
    Name:  azdevops-deployment
  Triggers:
    Authentication Ref:
      Name:  pipeline-trigger-auth
    Metadata:
      Organization URL From Env:  AZP_URL
      Pool ID:                    1
    Type:                         azure-pipelines
Status:
  Conditions:
    Message:  ScaledObject is defined correctly and is ready for scaling
    Reason:   ScaledObjectReady
    Status:   True
    Type:     Ready
    Message:  Scaling is not performed because triggers are not active
    Reason:   ScalerNotActive
    Status:   False
    Type:     Active
    Message:  No fallbacks are active on this scaled object
    Reason:   NoFallbackFound
    Status:   False
    Type:     Fallback
  External Metric Names:
    azure-pipelines-queue-xyzweb-1
  Health:
    azure-pipelines-queue-xyzweb-1:
      Number Of Failures:  0
      Status:              Happy
  Original Replica Count:  1
  Scale Target GVKR:
    Group:            apps
    Kind:             Deployment
    Resource:         deployments
    Version:          v1
  Scale Target Kind:  apps/v1.Deployment
Events:
  Type    Reason              Age   From           Message
  ----    ------              ----  ----           -------
  Normal  KEDAScalersStarted  47m   keda-operator  Started scalers watch
```

**Reply by SandipGorde-TomTom** [reply]
Sorry for the wrong formatting

**Reply by tomkerkhove** [reply]
And what are you seeing? Pending builds without the agent scaling?

**Reply by SandipGorde-TomTom** [reply]
> 

yes, autoscaling is not happening as the azure pipeline jobs are still in the queue and the expected behavior is that it should scale the pods of the deployment eventually increasing number of agents in the pipeline

**Reply by tomkerkhove** [reply]
Is it allowed to post a screenshot + URL of your pool in Azure Pipelines?

**Reply by SandipGorde-TomTom** [reply]
![image](https://user-images.githubusercontent.com/86930209/143259638-2c185a31-9ca3-4e7d-85ca-5fab901f7b01.png)

**Reply by tomkerkhove** [reply]
Sorry, I meant the dedicated page for your agent pool

**Reply by SandipGorde-TomTom** [reply]
it is also normal that the field ACTIVE is showing as FALSE ?
![image](https://user-images.githubusercontent.com/86930209/143259999-4cd78013-05ef-4e8b-85c6-7fa464eb9fd3.png)

**Reply by SandipGorde-TomTom** [reply]
> 

Unfortunately against policy, sorry. If you need more information from the configuration please let me know as I have been trying to troubleshoot this since yesterday

**Reply by tomkerkhove** [reply]
As you can see above, Active means that it's scaling which is not the case.

Are you allowed to share privately? If so, can you [send it](https://www.dropbox.com/request/s5y3JQ33IVCPvMvoMKlU) to me please?

**Reply by SandipGorde-TomTom** [reply]
Let me know what you want to check exactly, I will verify an let you know as I see one agent in the azure devops private agent pool. Whenever there is a queue of jobs the number of agents should scale using KEDA. Does not KEDA show any issues if there is anything wrong ? for example wrong URL, PAT etc?

**Reply by zroubalik** [reply]
There are errors that you'd see in the KEDA Operator logs for sure: https://github.com/kedacore/keda/blob/dbdba1e6b685e22e3786147f34203ebe4514488a/pkg/scalers/azure_pipelines_scaler.go#L136-L153

**Reply by tomkerkhove** [reply]
A screenshot of your Azure Pipelines' agent pool overview would be nice, thank you!

Screenshots from these would be nice:
- `https://dev.azure.com/{org-name}/{project-name}/_settings/agentqueues?queueId=10&view=jobs`
- `https://dev.azure.com/{org-name}/{project-name}/_settings/agentqueues?queueId=10&view=agents`
- `https://dev.azure.com/{org-name}/{project-name}/_settings/agentqueues?queueId=10&view=analytics`

**Reply by SandipGorde-TomTom** [reply]
> 

I hope you can understand, I tried to take the screenshot using my paint skill :D  Sorry for that

![tempsnip](https://user-images.githubusercontent.com/86930209/143263017-78822861-2915-4a14-af79-1d7ac67131df.png)

**Reply by SandipGorde-TomTom** [reply]
Can we go back and see what is the issue from the KEDA side ?

**Reply by SandipGorde-TomTom** [reply]
![tempsnip1](https://user-images.githubusercontent.com/86930209/143264342-73f88e2e-2b06-4b10-b50a-a876825e82e0.png)

**Reply by tomkerkhove** [reply]
Can you add the analytics page as well please?

**Reply by tomkerkhove** [reply]
> Can we go back and see what is the issue from the KEDA side ?



Trying to understand what's wrong :)

**Reply by SandipGorde-TomTom** [reply]
> 

![tempsnip2](https://user-images.githubusercontent.com/86930209/143265770-be870673-ce9b-4220-9c92-ce8d2d67d57f.png)

**Reply by tomkerkhove** [reply]
Hm doesn't this show that the agent was running the pipeline along with the screenshot above?

**Reply by SandipGorde-TomTom** [reply]
In the first screenshot where agent is being shown, whenever there is any running job it shows that job is running on this agent

**Reply by SandipGorde-TomTom** [reply]
![tempsnip3](https://user-images.githubusercontent.com/86930209/143268067-283a97a9-b294-4d48-89ae-60b0794b45b3.png)
I just ran 2 jobs together which  you can see in the image that one job is running while another is in pending state(second row)

**Reply by tomkerkhove** [reply]
Any changes inside the cluster? KEDA Logs/Scaled Object

**Reply by SandipGorde-TomTom** [reply]
below are metric logs:

```
Trace[1149509107]: [667.336006ms] [667.336006ms] END
I1124 15:20:37.012783       1 trace.go:205] Trace[402107940]: "List" url:/apis/external.metrics.k8s.io/v1beta1/namespaces/ado-agent/azure-pipelines-queue-xyzweb-1,user-agent:kube-controller-manager/v1.21.2 (linux/amd64) kubernetes/8a4e27b/system:serviceaccount:kube-system:horizontal-pod-autoscaler,client:10.0.40.84 (24-Nov-2021 15:20:36.365) (total time: 646ms):
Trace[402107940]: ---"Listing from storage done" 646ms (15:20:00.012)
Trace[402107940]: [646.992453ms] [646.992453ms] END
I1124 15:27:22.127497       1 trace.go:205] Trace[933967425]: "List" url:/apis/external.metrics.k8s.io/v1beta1/namespaces/ado-agent/azure-pipelines-queue-xyzweb-1,user-agent:kube-controller-manager/v1.21.2 (linux/amd64) kubernetes/8a4e27b/system:serviceaccount:kube-system:horizontal-pod-autoscaler,client:10.0.40.84 (24-Nov-2021 15:27:21.400) (total time: 726ms):
Trace[933967425]: ---"Listing from storage done" 726ms (15:27:00.127)
Trace[933967425]: [726.524586ms] [726.524586ms] END
I1124 15:27:37.245058       1 trace.go:205] Trace[1342381351]: "List" url:/apis/external.metrics.k8s.io/v1beta1/namespaces/ado-agent/azure-pipelines-queue-xyzweb-1,user-agent:kube-controller-manager/v1.21.2 (linux/amd64) kubernetes/8a4e27b/system:serviceaccount:kube-system:horizontal-pod-autoscaler,client:10.0.40.84 (24-Nov-2021 15:27:36.379) (total time: 865ms):
Trace[1342381351]: ---"Listing from storage done" 865ms (15:27:00.244)
Trace[1342381351]: [865.549655ms] [865.549655ms] END
I1124 15:29:52.072466       1 trace.go:205] Trace[1339189757]: "List" url:/apis/external.metrics.k8s.io/v1beta1/namespaces/ado-agent/azure-pipelines-queue-xyzweb-1,user-agent:kube-controller-manager/v1.21.2 (linux/amd64) kubernetes/8a4e27b/system:serviceaccount:kube-system:horizontal-pod-autoscaler,client:10.0.40.84 (24-Nov-2021 15:29:51.402) (total time: 670ms):
Trace[1339189757]: ---"Listing from storage done" 670ms (15:29:00.072)
Trace[1339189757]: [670.292614ms] [670.292614ms] END
```

**Reply by SandipGorde-TomTom** [reply]
How do we get logs of scaledobject? I could only describe it

**Reply by SandipGorde-TomTom** [reply]
do you think there could be any issue with the azure devops agent deployment k8s file ? Is it working for you with aws + ado agent ?

**Reply by SandipGorde-TomTom** [reply]
Let me know if you want to troubleshoot together

**Reply by JorTurFer** [reply]
Hi, I have tried it recently in [this context](https://github.com/kedacore/keda/issues/2293) and it works.

**Reply by JorTurFer** [reply]
![image](https://user-images.githubusercontent.com/36899226/143323608-b2461aae-4059-4503-8dd5-c79d1729d1e0.png)
![image](https://user-images.githubusercontent.com/36899226/143323603-104ec179-4238-4989-80b2-259f239bffe8.png)

**Reply by JorTurFer** [reply]
how did you get the poolId? did you get it using the comment? 
`Can be retrieved by the REST API call https://dev.azure.com/{organizationName}/_apis/distributedtask/pools?poolname={agentPoolName}`

**Reply by JorTurFer** [reply]
Maybe we should explain better in the docs how to recover the id 🤔

**Reply by tomkerkhove** [reply]
That's why I was asking for screenshots and URLs but that was not possible.

Let's better document it indeed, but normally if you go to the agent pool itself you should find the info there as well.

![image](https://user-images.githubusercontent.com/4345663/143386173-5541a715-03a4-4c2e-b5fe-f643e8ad0016.png)

But based on this support case, it seems like that is not always correct - Can you verify this please @SandipGorde-TomTom?

We should also improve the logs #2339

**Reply by SandipGorde-TomTom** [reply]
> 

yes can be retrieved by rest api, but how to retrieve, whether you will have enough permissions to fetch the pool id by rest api, if not then what is the alternative ? all these should be documented

**Reply by SandipGorde-TomTom** [reply]
> 

Poolid and queueid are two different things, I tried with the queueid and it did not work. Hence I fetched pool id by going to Organization -> settings> agent pool -> click on the agent pool name and then we will see poolid in the url. See below image for the reference
![tempsnip4](https://user-images.githubusercontent.com/86930209/143407717-da012307-e04c-4d66-9f68-67aec6479013.png)

**Reply by tomkerkhove** [reply]
The REST API call is documented on the YAML spec as per https://keda.sh/docs/2.4/scalers/azure-pipelines/ but we'll improve the docs to add it to the parameter list as well.

Ok thanks for confirming. We'll see how we can improve the logs more and throw exceptions if it was not found #2342

**Reply by SandipGorde-TomTom** [reply]
> The REST API call is documented on the YAML spec as per https://keda.sh/docs/2.4/scalers/azure-pipelines/ but we'll improve the docs to add it to the parameter list as well.
> 
> Ok thanks for confirming. We'll see how we can improve the logs more and throw exceptions if it was not found #2342

To be honest API did not work for me, however it is good to add more information and alternatives as well
